examples/composing-sandbox-nw-policies/README.md

```diff
@@ -0,0 +1,110 @@
+# Composing Sandbox with Network Policies
+
+The `Sandbox` API is a low-level primitive for creating secure sandboxes. In real-world scenarios, you often want to compose a `Sandbox` with other Kubernetes resources, such as `NetworkPolicy`, `Ingress` and `Service`, to create a more complete and secure "agentic sandbox" environment.
+
+* `Ingress` can be used to expose the Sandbox to Users and Agents outside the cluster. 
+* `NetworkPolicy` can be used to control who can connect to the Sandbox and limit the Sandbox outgoing connections to other pods or the internet.  
+* `Service` can be used to expose the Sandbox ports for `Ingress`  
+
+```mermaid
+graph TD
+    User --> Ingress
+    Agent --> Ingress
+    Ingress --> Service
+    Service --> NetworkPolicy
+    NetworkPolicy --> Sandbox
+    Sandbox --> NetworkPolicy
+    NetworkPolicy <--> Internet
+    NetworkPolicy <-->  Pods
+```
+
+This composition can be achieved in multiple ways:
+
+*  **Writing a custom controller:** You can write a dedicated Kubernetes controller that watches for a higher-level CRD and creates the `Sandbox`, `NetworkPolicy`, and `Service` resources.
+*  **Using a tool like KRO:** [KRO (Kubernetes Resource Orchestrator)](https://kro.run/docs/overview) allows you to define composite resources declaratively using a `ResourceGraphDefinition` (RGD). This creates a server side CRD that can be consumed by users.
+* **Using Helm:** Similar to KRO, we can use Helm to create a client side composition of resources and use the helm-cli to manage it.
+
+This example demonstrates the second approach, using KRO to define a new `AgenticSandbox` CRD that encapsulates a `Sandbox`, a `NetworkPolicy`, and a `Service`.
+
+The `Service` created here is distinct from the one created by the `agent-sandbox` controller. This layering allows for more complex networking configurations and showcases the value of composing resources.
+
+## Brief introduction to KRO
+
+KRO is a powerful tool for defining and managing composite resources in Kubernetes. At its core, KRO introduces the `ResourceGraphDefinition` (RGD) custom resource. An RGD allows you to declaratively specify a graph of Kubernetes resources that should be created, updated, and deleted together as a single unit. When an RGD is applied to a cluster, KRO automatically generates a new Custom Resource Definition (CRD) based on the RGD's schema. Users can then interact with this new, higher-level CRD, and KRO will handle the orchestration of the underlying resources defined in the RGD. This enables administrators to define complex, opinionated abstractions that users can consume easily, without needing to manage the individual components.
```

Comment:
Does this mean that when a Sandbox resource is deleted, it will also delete the NetworkPolicy object associated with it?

Reply:
The use deals with `AgenticSandbox` which creates the `Sandbox` and `NetworkPolicy`. Deleting the `AgenticSandbox` would delete both of them.